header/p2p/session.go

```diff
@@ -0,0 +1,224 @@
+	}
+}
+
+// doRequest chooses the best peer to fetch headers and sends a request in range of available maxRetryAttempts
+func (s *session) doRequest(
+	ctx context.Context,
+	stat *peerStat,
+	req *p2p_pb.ExtendedHeaderRequest,
+	headers chan []*header.ExtendedHeader,
+) {
+	select {
+	case <-ctx.Done():
+		return
+	case <-s.ctx.Done():
+		return
+	default:
+	}
+	r, size, duration, err := s.requestHeaders(ctx, stat.peerID, req)
+	if err != nil {
+		if err == context.Canceled {
+			return
+		}
+		log.Errorw("requesting headers from peer failed."+
+			"Retrying the request from different peer", "failed peer", stat.peerID, "err", err)
+		s.reqCh <- req
+		return
+	}
+
+	h, err := s.processResponse(r)
+	if err != nil {
+		s.errCh <- err
+		return
+	}
+	log.Debugw("request headers from peer succeed", "from", s.host.ID(), "pID", stat.peerID, "amount", req.Amount)
+	// send headers to the channel, update peer stats and return peer to the queue, so it can be re-used in case
+	// if there are other requests awaiting
+	headers <- h
+	stat.updateStats(size, duration)
+	s.queue.push(stat)
+}
+
+// handleOutgoingRequest pops peer from the queue and sends a prepared request to the peer.
+func (s *session) handleOutgoingRequest(ctx context.Context, result chan []*header.ExtendedHeader) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-s.ctx.Done():
+			return
+		case req := <-s.reqCh:
+			stats := s.queue.calculateBestPeer(ctx)
+			go s.doRequest(ctx, stats, req, result)
+		}
+	}
+}
+
+// GetRangeByHeight requests headers from different peers.
+func (s *session) getRangeByHeight(ctx context.Context, from, amount uint64) ([]*header.ExtendedHeader, error) {
+	log.Debugw("requesting headers", "from", from, "to", from+amount)
+	requests := prepareRequests(from, amount, headersPerPeer)
+	result := make(chan []*header.ExtendedHeader, len(requests))
+	headers := make([]*header.ExtendedHeader, 0, amount)
+	s.reqCh = make(chan *p2p_pb.ExtendedHeaderRequest, len(requests))
+
+	go s.handleOutgoingRequest(ctx, result)
+	for _, req := range requests {
+		s.reqCh <- req
+	}
+
+	for i := 0; i < cap(result); i++ {
+		select {
+		case <-s.ctx.Done():
+			return nil, errors.New("header/p2p: session was stopped")
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		case err := <-s.errCh:
+			return nil, err
+		case res := <-result:
+			headers = append(headers, res...)
+		}
+	}
+	s.cancel()
+	sort.Slice(headers, func(i, j int) bool {
+		return headers[i].Height < headers[j].Height
+	})
+	return headers, nil
+}
+
+// requestHeaders sends the ExtendedHeaderRequest to a remote peer.
+func (s *session) requestHeaders(
+	ctx context.Context,
+	to peer.ID,
+	req *p2p_pb.ExtendedHeaderRequest,
+) ([]*p2p_pb.ExtendedHeaderResponse, uint64, uint64, error) {
+	startTime := time.Now()
+	stream, err := s.host.NewStream(ctx, to, s.protocolID)
+	if err != nil {
+		return nil, 0, 0, err
+	}
+	if err = stream.SetWriteDeadline(time.Now().Add(writeDeadline)); err != nil {
```

Comment:
🚓 
Context governs the execution of requests, so the deadline should be taken from it. This is different from FraudSync. It is a background task; here, we have a client that anyone can use and pass their own context.

Reply:
What do you mean? we are using deadlines in Head:
[1](https://github.com/celestiaorg/celestia-node/blob/06b9c0c1f79c8114b9e0ab4661c7614af60aad7f/header/p2p/exchange.go#L187)
[2](https://github.com/celestiaorg/celestia-node/blob/06b9c0c1f79c8114b9e0ab4661c7614af60aad7f/header/p2p/exchange.go#L204)